route_guide/route_guide_server.cc

```diff
@@ -55,166 +66,259 @@ float GetDistance(const Point& start, const Point& end) {
+                       call.Write(f);
+                     }
+                   }
+                   // When writes are done just finish the call.
+                   call.Finish(Status::OK);
+                 });
+           }))
+        | Loop();
+  }
 
-    status = server->Serve<RouteGuide, Stream<Point>, RouteSummary>(
-        "RecordRoute",
-        [this](auto&& call) {
-          int point_count = 0;
-          int feature_count = 0;
-          float distance = 0.0;
-          Point previous;
-
-          system_clock::time_point start_time = system_clock::now();
-
-          call->OnRead(
-              [this, point_count, feature_count, distance, previous, start_time](
-                  auto* call, auto&& point) mutable {
-            if (point) {
-              point_count++;
-              if (!GetFeatureName(*point, feature_list_).empty()) {
-                feature_count++;
-              }
-              if (point_count != 1) {
-                distance += GetDistance(previous, *point);
-              }
-              previous = *point;
-            } else {
-              system_clock::time_point end_time = system_clock::now();
-              RouteSummary summary;
-              summary.set_point_count(point_count);
-              summary.set_feature_count(feature_count);
-              summary.set_distance(static_cast<long>(distance));
-              auto secs = std::chrono::duration_cast<std::chrono::seconds>(
-                  end_time - start_time);
-              summary.set_elapsed_time(secs.count());
-              call->WriteAndFinish(summary, Status::OK);
-            }
-          });
-        });
-
-    if (!status.ok()) {
-      return status;
-    }
+  // ServeRecordRoute() demonstrates a client streaming RPC. The  client
+  // sends some number of Point messages to the server. The server
+  // responds with a single message which contains the number of points,
+  // number of features found in the db, the resulting distance calculated from
+  // all Point coordinates, and the resulting time.
+  auto ServeRecordRoute(Server* server) {
+    return server->Accept<
+               RouteGuide,
+               Stream<Point>,
+               RouteSummary>("RecordRoute")
+        | Map(Then([this](auto&& context) {
+             int point_count = 0;
+             int feature_count = 0;
+             float distance = 0.f;
+             Point previous;
+             system_clock::time_point start_time = system_clock::now();
+             return Server::Handler(std::move(context))
+                 .body([this,
+                        point_count,
+                        feature_count,
+                        distance,
+                        previous,
+                        start_time](
+                           auto& call,
+                           auto&& point) mutable {
+                   // Check if the request is valid.
+                   if (point) {
```

Comment:
As with the comment below about type-safe checks:
```suggestion
                   if (point != nullptr) {
```